### PR TITLE
Shopping list

### PR DESCRIPTION
**Pull request by KhonkoDmytro**

### CODE REVIEW

**Review by forestik — APPROVED** [review]

File: core/src/main/java/greencity/controller/HabitController.java [hunk 1]
```diff
@@ -72,4 +70,25 @@
...
+     * @return List of {@link GoalDto}.
+     */
+    @ApiOperation(value = "Get shopping list.")
+    @ApiResponses(value = {
+        @ApiResponse(code = 200, message = HttpStatuses.OK),
+        @ApiResponse(code = 400, message = HttpStatuses.BAD_REQUEST),
+    })
+    @GetMapping("/shopping-list")
+    @ApiLocale
+    public ResponseEntity<List<GoalDto>> getGoals(
```
Comment (vterlyha) [inline core/src/main/java/greencity/controller/HabitController.java:88]: getGoals ?

File: core/src/main/java/greencity/controller/UserController.java [hunk 2]
```diff
@@ -218,44 +214,90 @@
     }
 
     /**
-     * Method returns list of user goals for specific language.
+     * Method returns list user custom goals.
```
Comment (vterlyha) [inline core/src/main/java/greencity/controller/UserController.java:217]: Create Goals controller and move everything relates to goals to it

File: dao/src/main/java/greencity/entity/HabitGoal.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        HabitGoal habitGoal = (HabitGoal) o;
+        return id.equals(habitGoal.id)
+            && habit.equals(habitGoal.habit)
+            && Objects.equals(goal, habitGoal.goal);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, habit, goal);
+    }
```
Comment (vterlyha) [inline dao/src/main/java/greencity/entity/HabitGoal.java:47]: Why don't you use Lombok ?

File: service/src/main/java/greencity/service/GoalServiceImpl.java [hunk 4]
```diff
@@ -218,4 +228,146 @@ private void setValueIfNotEmpty(List<SearchCriteria> searchCriteria, String key,
...
+        Optional<HabitAssign> habitAssignOptional =
+            habitAssignRepo.findByHabitIdAndUserIdAndSuspendedFalse(habitId, userId);
+        if (habitAssignOptional.isPresent()) {
+            HabitAssign habitAssign = habitAssignOptional.get();
+            if (goalDtos != null) {
+                saveGoalForUserGoal(habitAssign, goalDtos);
+            }
+        } else {
+            throw new NotFoundException(ErrorMessage.HABIT_ASSIGN_NOT_FOUND_BY_ID);
+        }
```
Comment (vterlyha) [inline service/src/main/java/greencity/service/GoalServiceImpl.java:248]: Move to private method

File: service/src/main/java/greencity/service/GoalServiceImpl.java [hunk 5]
```diff
@@ -218,4 +228,146 @@ private void setValueIfNotEmpty(List<SearchCriteria> searchCriteria, String key,
...
+                    userGoal.setHabitAssign(habitAssign);
+                    habitAssign.getUserGoals().add(userGoal);
+                    userGoalRepo.saveAll(habitAssign.getUserGoals());
+                } else {
+                    throw new WrongIdException(ErrorMessage.GOAL_ALREADY_SELECTED + el.getId());
+                }
+            } else {
+                throw new NotFoundException(ErrorMessage.GOAL_NOT_ASSIGNED_FOR_THIS_HABIT + el.getId());
+            }
+        }
```
Comment (vterlyha) [inline service/src/main/java/greencity/service/GoalServiceImpl.java:274]: Move to private methods

File: service/src/main/java/greencity/service/GoalServiceImpl.java [hunk 6]
```diff
@@ -218,4 +228,146 @@ private void setValueIfNotEmpty(List<SearchCriteria> searchCriteria, String key,
...
+                .map(userGoal -> modelMapper.map(userGoal, UserGoalResponseDto.class))
+                .collect(Collectors.toList());
+            if (goalDtos.isEmpty()) {
+                throw new UserHasNoGoalsException(ErrorMessage.USER_HAS_NO_GOALS);
+            }
+            goalDtos.forEach(el -> setTextForUserGoal(el, language));
+            return goalDtos;
+        } else {
+            throw new UserHasNoGoalsException(ErrorMessage.USER_HAS_NO_GOALS);
+        }
```
Comment (vterlyha) [inline service/src/main/java/greencity/service/GoalServiceImpl.java:299]: Same here

File: service/src/main/java/greencity/service/GoalServiceImpl.java [hunk 7]
```diff
@@ -218,4 +228,146 @@ private void setValueIfNotEmpty(List<SearchCriteria> searchCriteria, String key,
...
+    public UserGoalResponseDto updateUserGoalStatus(Long userId, Long goalId, String language) {
+        UserGoal userGoal;
+        userGoal = userGoalRepo.getOne(goalId);
+        if (GoalStatus.ACTIVE.equals(userGoal.getStatus())) {
+            userGoal.setStatus(GoalStatus.DONE);
+            userGoal.setDateCompleted(LocalDateTime.now());
+            userGoalRepo.save(userGoal);
+        } else {
+            throw new UserGoalStatusNotUpdatedException(ErrorMessage.USER_GOAL_STATUS_IS_ALREADY_DONE + goalId);
+        }
```
Comment (vterlyha) [inline service/src/main/java/greencity/service/GoalServiceImpl.java:335]: and here

**Review by Markiyan000 — APPROVED** [review]

**Review by vterlyha — APPROVED** [review]

**Pull request merged**